### PR TITLE
feat(dev): add publish-artifact skill

### PR DESCRIPTION
**Pull request by edruder**
## Summary

Adds a new **`publish-artifact`** skill to the `rp1-dev` plugin. It publishes an rp1 work artifact (investigation report, design doc, audit) as an **idempotent PR or issue comment** — keyed by an HTML marker so re-runs update the same comment in place — **without committing the artifact file to the repo**. The artifact is read-only locally and write-only on the GitHub side.

Invoked as `/rp1-dev:publish-artifact <artifact-path> [target] [--dry-run] [--force]`.

This was developed as a standalone plugin ([`anvilco/persist-rp1-artifact`](https://github.com/anvilco/persist-rp1-artifact)) and is now being folded permanently into rp1, so it ships to `rp1-dev` users alongside the other PR-workflow skills (`pr-review`, `address-pr-feedback`, `pr-walkthrough`, `pr-visual`, `pr-stack`).

### Adaptation to rp1 conventions
- **Renamed** `persist-rp1-artifact` → `publish-artifact` — the `/rp1-dev:` namespace already supplies the `rp1` qualifier, matching the short, unprefixed names of sibling skills.
- **Frontmatter** converted to the canonical rp1 SKILL.md format (`metadata` block: `category: review`, tags, `allowed-tools`, structured `arguments`). The build pipeline auto-derives `argument-hint` and injects the `## 0. Resolve Arguments` section.
- **Dropped** standalone-repo scaffolding (`plugin.json`, `marketplace.json`, `LICENSE`, `README`, `screenshots/`) and the donor git history — this lives in rp1 now.
- **Neutralized** org-specific example URLs in tests (`anvilco/pdf-service` → `rp1-run/rp1`).
- **Regenerated** the guide `CATALOG.md` and updated the `rp1-dev` README command counts.

### Note for reviewers
The helper scripts are **Python** (`parse_target.py`, `project.py`, `publish.py`) with a 60-test `unittest` suite. This is an outlier in an otherwise Bun/TypeScript repo; happy to file or take a follow-up to port them to Bun/TS. Flagging here so it is a conscious decision rather than an oversight.

## Test Plan

- `python3 -m unittest discover -s plugins/dev/skills/publish-artifact/tests` → **60 passed** (byte-exact golden fixtures for `project.py` + orchestration unit tests for `publish.py` with `gh` never called).
- `bun run scripts/build-claude-code.ts` → builds clean; `rp1-dev` reports 23 skills; `argument-hint` and the Resolve Arguments section generate correctly; `CATALOG.md` picks up the new skill.

## Related Issues

None.

### CODE REVIEW

**Comment by cloud-on-prem** [comment]
Thanks for this @edruder - and for proactively flagging the Python yourself. 
The skill itself is genuinely well-built (the dry-run-first safety, the idempotent marker, the refuse-on-foreign/duplicate logic). 
Digging in, I think the Python scripts and the "too many examples" are actually the same issue with a single fix: the examples/ files are really byte-exact golden fixtures for `project.py`, and the helper logic wants to live in rp1 agent-tools (TypeScript) rather than skill-local Python. 
The github-pr agent-tool already has the Octokit comment plumbing, so folding the projection + idempotent upsert there drops the python3 dependency entirely, moves the 60 tests (we can probably cut down on the number of tests as well) into our standard Bun test tree, and relocates the fixtures to the usual golden/ dir — which collapses the skill back to a normal SKILL.md + references/. 

I'd lean toward doing this in this PR rather than as a follow-up, mainly because the build currently ships the Python scripts + tests + fixtures to every rp1-dev user and python3 isn't a guaranteed 100% runtime dependency - although most systems have it. I'm attaching a note with a concrete file-by-file mapping (and the two minor bugs I'd fix during the port) as a possible path - happy to talk through it.

❤️ 

P.S: If you are happy with the proposal below, feel free to point this to rp1.

```sh
/rp1-dev:build-fast publish-artifact
  context: <this-github-comment-url>
```

----

# publish-artifact: fold Python helpers into `rp1 agent-tools`

Date: 2026-06-06

Context: review of [PR #397](https://github.com/rp1-run/rp1/pull/397) (`feat(dev): add publish-artifact skill`).

## Summary

PR #397 ships a `publish-artifact` skill whose logic lives in three skill-local **Python**
scripts (`parse_target.py`, `project.py`, `publish.py`) with a co-located Python `tests/` dir and
an `examples/` dir of 16 byte-exact golden fixtures.

The proper fix is **one move**: port the logic into `rp1 agent-tools` as TypeScript, reusing the
existing `github-pr` tool. The skill then calls `rp1 agent-tools github-pr publish-artifact …`
(or a sibling op), the Python disappears, the tests become Bun tests under `cli/src/__tests__/`,
and the golden fixtures move to a standard `golden/` dir. This resolves **both** reviewer concerns
("too many examples", "Python outlier") at once, because both are downstream of the same decision
to put procedural logic in skill-local scripts.

## Problem (why skill-local Python doesn't fit)

Short version — evidence is in the review report:

- **New runtime dependency.** `python3` is declared nowhere in the repo and is not part of the
  single-executable Bun CLI. A user without `python3` on PATH gets a broken skill. The only other
  skill with a `scripts/` dir (`utils/prompt-eval-builder`) uses `#!/usr/bin/env bun`.
- **No path-resolution mechanism.** `python3 "$SKILL_DIR/scripts/publish.py"` assumes the skill can
  learn its own install path. No such mechanism exists in rp1; it's fragile across Claude Code /
  OpenCode / Codex.
- **Build ships everything.** `cli/src/build/parser.ts` copies every skill subdir file verbatim
  (no filter), so `tests/` (Python unittest) and `examples/` (fixtures) ship to every rp1-dev user.
- **Wrong home for logic.** The established pattern is `rp1 agent-tools <subcommand>` — deterministic
  TypeScript compiled into the CLI. `github-pr` already owns the GitHub comment surface.

## Target end state

```
plugins/dev/skills/publish-artifact/
  SKILL.md            # frontmatter + procedure that calls `rp1 agent-tools …` + reference pointers
  references/         # KEEP — artifact-frontmatter.md, projection-format.md, edge-cases.md
                      #         (progressive-disclosure pattern, matches prompt-writer exemplar)
  # scripts/  → gone (logic moved to cli/src/agent-tools/github-pr/)
  # tests/    → gone (moved to cli/src/__tests__/agent-tools/github-pr/)
  # examples/ → gone (golden fixtures moved to cli/src/__tests__/.../golden/)
```

Result: skill collapses to the normal `SKILL.md` (+ optional `references/`) shape — no longer the
only skill with 4 subdirs, no shipped Python, no shipped fixtures.

## Proposed design

### 1. New agent-tool operation on `github-pr`

`github-pr` already has the plumbing this needs:

```
cli/src/agent-tools/github-pr/
  client.ts        # Octokit client (reuse)
  validation.ts    # input validation (extend)
  models.ts        # input/output types (extend)
  fetch-comments.ts # already lists PR/issue comments (reuse for the marker lookup)
  submit-review.ts  add-reaction.ts  reply-comment.ts
  index.ts         # exports execute* fns; registration
```

None of the four existing ops does an **idempotent top-level issue comment upsert** (submit-review
posts a *review*; reply-comment replies to a *review thread*). So add one new operation:

- New file `cli/src/agent-tools/github-pr/upsert-comment.ts` exporting `executeUpsertComment`.
- Add types to `models.ts`, validation to `validation.ts`, export from `index.ts`, wire the
  subcommand in `cli/src/agent-tools/command.ts` (alongside the existing `github-pr` subcommands
  registered around `command.ts:149`).
- Use **Octokit** (consistent with the rest of `github-pr`), not the `gh` CLI subprocess that
  `publish.py` shells out to. Octokit calls: `issues.listComments` (lookup),
  `issues.createComment` (POST), `issues.updateComment` (PATCH), `users.getAuthenticated` (owner
  check), `pulls.get` / `issues.get` (state + PR-vs-issue probe).

Suggested input contract (stdin JSON, matching `fetch-comments` style):

```json
{
  "artifact_path": ".rp1/work/.../report.md",
  "target": "397 | https://github.com/o/r/pull/397 | <omitted = current branch PR>",
  "dry_run": false,
  "force": false
}
```

Output JSON: `{ action: "POST"|"PATCH"|"would-POST"|"would-PATCH", comment_url, doc_key, size_bytes, warnings[] }`.

### 2. Projection logic → its own pure TS module + golden tests

`project.py` is pure (no network) — the easy, high-value port. Put it in a dedicated module so it
is unit-testable in isolation, e.g. `cli/src/agent-tools/github-pr/artifact-projection.ts`:

- `splitFrontmatter`, `deriveTitle`, `stripLeadingH1`, `extractSummary` (the deterministic ladder,
  rungs 0–6), `buildTableRows`, `markerKey`, `assemble`, `checkSize`.
- Golden tests: render each `*-input.md` fixture, compare byte-for-byte to `*-output.md`, under
  `cli/src/__tests__/agent-tools/github-pr/golden/` (mirrors the existing
  `cli/src/__tests__/build/templates/golden/` + `readGolden()` pattern). Fixtures move here from
  the skill's `examples/`; they are **test data**, never shipped.

### 3. Orchestration / idempotency

`publish.py`'s pure decision functions port directly and get Bun unit tests (gh/Octokit mocked,
matching `test_publish.py`'s "never call the network" design):

- `findMarkerMatches`, `decideAction` (0→POST, 1-mine→PATCH, 1-foreign→refuse-unless-force,
  ≥2→refuse-always), `softDetectOrphan`, `mtimeWarning`, `buildDiagnostic`, `formatSuccess`.
- Fix the two low-severity bugs noted in the review during the port:
  - **tz**: `mtime_warning` parsed GitHub's UTC timestamp as naive-local. Use a proper ISO/UTC
    `Date` in TS.
  - **line endings**: frontmatter regex assumed `\n`; tolerate CRLF.

### 4. SKILL.md after

Procedure section becomes one command, no `$SKILL_DIR`, no `python3`:

```bash
echo '{"artifact_path":"<path>","target":"<target>","dry_run":<bool>,"force":<bool>}' \
  | rp1 agent-tools github-pr publish-artifact
```

Drop `Bash(python3 *)` from `allowed-tools`; keep `Bash(rp1 *)`. `references/` stays as the spec.

## Python → TypeScript mapping

| PR #397 (Python, skill-local)            | Proposed home (TypeScript, in CLI)                                   |
|------------------------------------------|----------------------------------------------------------------------|
| `scripts/project.py` (pure projection)   | `cli/src/agent-tools/github-pr/artifact-projection.ts`               |
| `scripts/parse_target.py` (URL/number)   | helper in `upsert-comment.ts` or `validation.ts`                     |
| `scripts/publish.py` (orchestrator + gh) | `cli/src/agent-tools/github-pr/upsert-comment.ts` (Octokit)          |
| `tests/test_project.py` (golden)         | `cli/src/__tests__/agent-tools/github-pr/projection.test.ts`         |
| `tests/test_publish.py` (decision logic) | `cli/src/__tests__/agent-tools/github-pr/upsert-comment.test.ts`     |
| `examples/*-input.md` ↔ `*-output.md`    | `cli/src/__tests__/agent-tools/github-pr/golden/` (not shipped)      |
| `tests/test_parse_target.py`             | folded into the upsert-comment test                                  |

## Decisions / open questions

1. **New op vs. new tool.** Lean: add `publish-artifact` (or `upsert-comment`) **to `github-pr`** —
   reuses client/validation/fetch and keeps GitHub ops in one tool. Alternative: a standalone
   `rp1 agent-tools publish-artifact` if we want projection reusable outside GitHub.
2. **Octokit vs `gh`.** Lean Octokit for consistency + testability. Confirm token resolution
   matches how `client.ts` already authenticates in CI and locally.
3. **Scope for this PR.** The logic maps 1:1 onto existing structure, so the port is the
   lower-total-cost path. If a same-PR rewrite is too large, the interim must at least stop
   shipping Python + fixtures to users — but a build-exclude/`python3`-preflight only papers over
   the runtime-dependency and path-resolution issues; prefer the port.

## Why this resolves both reviewer concerns

- *"Too many examples"* → the 16 files were golden fixtures, not agent examples. Once projection is
  a TS module tested under `__tests__/.../golden/`, they leave the skill entirely.
- *"Python scripts — match the pattern, or fold into agent-tools?"* → they don't match the pattern;
  folding into `github-pr` is the pattern. No `python3`, no `$SKILL_DIR`, no shipped tests.

One decision (logic belongs in `rp1 agent-tools`) collapses 4 subdirs and ~1.6k lines of
skill-local Python+fixtures down to a TS op + standard tests + a thin SKILL.md.

**Comment by edruder** [comment]
@cloud-on-prem, take a look at what Claude and I came up with!

**Comment by cloud-on-prem** [comment]
🤖 Pushed follow-up commits to this PR (also merges current `main`) addressing the review on the write path:

- **`github.com` host pinning** + **URL number boundary** so a non-github origin or a malformed `…/pull/123abc` can't mis-target a write (M2, M3)
- **CRLF normalization** in projection — a CRLF artifact no longer leaks `\r` into the comment body (M4)
- SKILL.md corrected to the real `{ success, tool, data: {…} }` envelope; dry-run documented as still running the safety gates (M1, M5)
- removed the unused dry-run/success formatter cluster (M6); made the **no-`gh`-subprocess** guard real (stubs `Bun.spawn`/`child_process`) and fixed the REQ-007 epoch comment (M7, M8)
- low-severity tidy-ups: warnings channel, orphan wording, README count, doc-key dedup

Verification: github-pr suite **83/83**, `just check-cli` + `just catalog-check` green. PR description left as-is.

---

@edruder - what do you think?

File: cli/src/__tests__/agent-tools/github-pr/publish-decisions.test.ts [hunk 1]
```diff
@@ -7,7 +7,8 @@
  *
  * A final block exercises `executePublishComment`'s create-vs-update branch
  * against a mocked Octokit to confirm POST calls `issues.createComment`, PATCH
- * calls `issues.updateComment`, and no `gh` subprocess is spawned (REQ-001).
+ * calls `issues.updateComment`, and — with Bun.spawn and child_process stubbed
+ * to throw — that no `gh` subprocess is spawned (REQ-001).
```
Comment (edruder) [inline cli/src/__tests__/agent-tools/github-pr/publish-decisions.test.ts]: nit: `that` is extraneous.

**Comment by edruder** [comment]
These changes all look good! Thanks for curating!

**Comment by edruder** [comment]
when using this skill recently, i realized that sometimes the top of the artifact document doesn't give enough useful signal. not because the artifact is wrong in any way, but simply because the "top" of the document isn't intended to be a TL;DR summary, suitable for "above the fold" treatment.

i'm thinking that after the base skill is merged, a follow-on PR could add an option to have AI summarize the artifact, and put that summary above the fold.
